### ISSUE TITLE
Implement HTMLScriptElement.type and similar attributes

### ISSUE THREAD

**Issue by emilio** [servo/servo#10227]
There are a few attribures missing in HTMLScriptElement that should be relatively easy to implement, and are currently missing.

I'm explicitely talking about https://github.com/servo/servo/blob/master/components/script/dom/webidls/HTMLScriptElement.webidl#L9-L13:

```
  //         attribute DOMString type;
  //         attribute DOMString charset;
  //         attribute boolean async;
  //         attribute boolean defer;
  //         attribute DOMString crossOrigin;
```

Some of them should be just reading attribute values if I'm not wrong (`crossOrigin` could be a bit trickier, I don't currently know the current CORS infrastructure).

@nox could be better suited to clear doubts regarding the implementation.

**Comment by jdm** [comment]
https://html.spec.whatwg.org/multipage/scripting.html#dom-script-type . 
I think `async` shouldn't be included here, since it's tied in with the rest of the recent specification changes to support script "modules".

**Comment by Shiroy** [comment]
I will work on this issue.

**Comment by roysohan** [comment]
How do I start contributing to this. I am a total beginner, do not know much, so what are the files I need to edit and what should I do, where should I start.

**Comment by jdm** [comment]
First step is to expose the attribute via the WebIDL file linked in the original description. Rebuilding will cause a new compile error to occur in `htmlscriptelement.rs`. For `type`, the goal is to avoid executing any script tag that has a type attribute that doesn't represent JavaScript (such as `text/javascript`). This logic exists in `htmlscriptelement.rs`.

**Comment by roysohan** [comment]
So can you please dumb it down. We can start one by one. First how do I expose the attribute in webidl file, by making it public? or some other modifier?

**Comment by jdm** [comment]
It's currently commented out, so by uncommenting it.

**Comment by roysohan** [comment]
Its asking whether to create a pull request, so should i create a PR?

**Comment by jdm** [comment]
Pull requests are only created when you have local code changes that you wish to have merged into the original repository (ie. this one).

**Comment by roysohan** [comment]
Created a PR, what to do next?

**Comment by jdm** [comment]
http://doc.servo.org/script/dom/bindings/index.html may help understand why there's a Rust compile error after modifying a webidl file.

**Comment by roysohan** [comment]
So it means I have to first work on the type attribute before making the compilation a success?

**Comment by roysohan** [comment]
Sir what to do next, I have forked it out and doing on my own fork, located htmlscriptelement.rs
so what to do now how to avoid those javascript types?

**Comment by jdm** [comment]
Take a look at the WebIDL file for other similar attributes. Find out how they're implemented in htmlscriptelement.rs already, and use them for inspiration!

**Comment by jdm** [comment]
For clarity, the goal is that running Servo on an HTML file with
```html
<script type="wheeeee">
definitely not javascript
</script>
```
will not report any errors in the terminal.

**Comment by roysohan** [comment]
the htmlscriptelement.rs file looks like a big soup. I am totally lost

**Comment by roysohan** [comment]
what does the get() function and @atom() do?

**Comment by jdm** [comment]
Sorry, `get` isn't descriptive enough for me to be able to answer that. `atom!` is just a way of turning a particular string into an `Atom` value (which is basically a kind of string that's really fast to compare). [The documentation](http://doc.servo.org/string_cache/atom/struct.Atom.html) doesn't describe it, unfortunately.

**Comment by roysohan** [comment]
So what should I do now, sir? I am a little confused. Iknow that it has some connection to step 6, but what to do next?

**Comment by roysohan** [comment]
I think I should leave this, since currently I do not have enough bandwidth to download servo and run it. I think before making any advances I should be able to compile it. And it is much complicated for me.

**Comment by Ms2ger** [comment]
You could try using https://janitor.technology

**Comment by roysohan** [comment]
I have already asked for an invitation, if I get an invitation then I will slowly contribute on this issue

**Comment by perlun** [comment]
I'll give this a look.

**Comment by perlun** [comment]
@jdm - FWIW, I don't get any errors even before implementing this on a file like this:

```html
<html>
<head>
<script foo="burp" type="wheeeee">
definitely not javascript
</script>
</head>
</html>
```

**Comment by jdm** [comment]
Huh, so `HTMLScriptElement::is_javascript` already exists.  I was going by https://github.com/servo/servo/pull/10224#issuecomment-202090141; @emilio could you elaborate on what wasn't working precisely?

@perlun Regardless, we can still work on exposing the relevant attributes appropriately :)

**Comment by perlun** [comment]
> Regardless, we can still work on exposing the relevant attributes appropriately :)

Yep. :smile: Hoping to have something coming up in a while.

**Comment by emilio** [comment]
@jdm it is common to test script.type when writing webgl stuff, to know the type of the shader.

**Comment by roysohan** [comment]
When we write <script type="wheeeee"> definitely not javascript </script>
What are the parts of <script type="wheeee"> and what are the variables
used for storing
What is the "wheeee" called and where it is stored

**Comment by KiChjang** [comment]
Fixed by #10354.
